src/nwb_conversion_tools/tools/hdmf.py

```diff
@@ -90,3 +91,66 @@ def _get_maxshape(self) -> tuple:
 
     def _get_data(self, selection: Tuple[slice]) -> np.ndarray:
         return self.data[selection]
+
+
+class ImagingExtractorDataChunkIterator(GenericDataChunkIterator):
+    """
+    Generic data chunk iterator for an ImagingExtractor object
+    primarily used when writing imaging data to an NWB file.
+    """
+
+    def __init__(
+        self,
+        imaging_extractor: ImagingExtractor,
+        buffer_gb: Optional[float] = None,
+        buffer_shape: Optional[tuple] = None,
+        chunk_mb: Optional[float] = None,
+        chunk_shape: Optional[tuple] = None,
+        display_progress: bool = False,
+        progress_bar_options: Optional[dict] = None,
+    ):
+        self.imaging_extractor = imaging_extractor
+
+        if buffer_gb is None and buffer_shape is None:
+            buffer_gb = 1.0
+
+        if buffer_shape is None:
+            buffer_shape = self._get_default_buffer_shape(buffer_gb=buffer_gb)
+
+        if chunk_mb is None and chunk_shape is None:
+            chunk_mb = 1.0
+
+        super().__init__(
+            buffer_shape=buffer_shape,
+            chunk_mb=chunk_mb,
+            chunk_shape=chunk_shape,
+            display_progress=display_progress,
+            progress_bar_options=progress_bar_options,
+        )
+
+        assert self.buffer_shape[1:] == self.maxshape[1:], (
+            f"Except from the first axis, the buffer shape ({self.buffer_shape}) "
+            f"must be equal to max shape ({self.maxshape})."
+        )
+
+    def _get_default_buffer_shape(self, buffer_gb: float) -> tuple:
+        num_frames_in_buffer_gb = int(
+            buffer_gb * 1e9 / (np.prod(self.imaging_extractor.get_image_size()) * self._get_dtype().itemsize)
+        )
```

Comment:
A better approach here would be to calculate the exact size of the chunk shape (same `np.prod` approach as above but using units of chunks instead of frames) and calculating how many chunks you can fit into the buffer.

Since frames is the only 'free' dimension here, the calculation is very similar but that will ensure that axis is a proper multiple (i.e., that the buffer frames divided by chunk frames is an integer)